File: Backend/src/main/java/com/solekta/solekta/service/ServiceService.java
package com.solekta.solekta.service;

import com.solekta.solekta.model.RentalService;
import com.solekta.solekta.repository.ServiceRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.multipart.MultipartFile;
import lombok.extern.slf4j.Slf4j;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

@Component
@Slf4j
public class ServiceService {

    @Autowired
    private ServiceRepository serviceRepository;

    // CREATE - Add new service
    public RentalService createService(RentalService service) {
        return serviceRepository.save(service);
    }

    public RentalService createServiceWithImage(String serviceName, String description,
            BigDecimal pricePerDay, BigDecimal pricePerHour, String category,
            Boolean isAvailable, Integer minRentalPeriod, Integer maxRentalPeriod,
            MultipartFile imageFile) {

        RentalService service = new RentalService();
        service.setServiceName(serviceName);
        service.setDescription(description);
        service.setPricePerDay(pricePerDay);
        service.setPricePerHour(pricePerHour);
        service.setCategory(category);
        service.setIsAvailable(isAvailable);
        service.setMinRentalPeriod(minRentalPeriod);
        service.setMaxRentalPeriod(maxRentalPeriod);
        
        // Handle image file upload
        if (imageFile != null && !imageFile.isEmpty()) {
            try {
                service.setImageData(imageFile.getBytes());
                service.setImageFileName(imageFile.getOriginalFilename());
                service.setImageContentType(imageFile.getContentType());
                log.info("Service image uploaded: {}", imageFile.getOriginalFilename());
            } catch (IOException e) {
                log.error("Failed to process service image file: {}", e.getMessage());
                throw new RuntimeException("Failed to process service image file", e);
            }
        }
        
        return serviceRepository.save(service);
    }

    // READ - Get all services
    public List<RentalService> getAllServices() {
        return serviceRepository.findAllOrderByName();
    }

    // READ - Get available services
    public List<RentalService> getAvailableServices() {
        return serviceRepository.findByIsAvailableTrue();
    }

    // READ - Get service by ID
    public Optional<RentalService> getServiceById(Long id) {
        return serviceRepository.findById(id);
    }

    // READ - Search services by keyword
    public List<RentalService> searchServices(String keyword) {
        return serviceRepository.searchServices(keyword);
    }

    // READ - Get services by category
    public List<RentalService> getServicesByCategory(String category) {
        return serviceRepository.findByCategoryContainingIgnoreCase(category);
    }

    // READ - Get services by price range
    public List<RentalService> getServicesByPriceRange(BigDecimal minPrice, BigDecimal maxPrice) {
        return serviceRepository.findByPricePerDayBetween(minPrice, maxPrice);
    }

    // UPDATE - Update existing service
    public RentalService updateService(Long id, RentalService updatedService) {
        return serviceRepository.findById(id)
                .map(service -> {
                    service.setServiceName(updatedService.getServiceName());
                    service.setDescription(updatedService.getDescription());
                    service.setPricePerDay(updatedService.getPricePerDay());
                    service.setPricePerHour(updatedService.getPricePerHour());
                    service.setCategory(updatedService.getCategory());
                    service.setIsAvailable(updatedService.getIsAvailable());
                    service.setMinRentalPeriod(updatedService.getMinRentalPeriod());
                    service.setMaxRentalPeriod(updatedService.getMaxRentalPeriod());
                    return serviceRepository.save(service);
                })
                .orElseThrow(() -> new RuntimeException("Service not found with id: " + id));
    }

    // DELETE - Delete service
    public void deleteService(Long id) {
        serviceRepository.deleteById(id);
    }

    // CHECK - Service exists
    public boolean serviceExists(Long id) {
        return serviceRepository.existsById(id);
    }

    // Helper method to save service
    public RentalService saveService(RentalService service) {
        return serviceRepository.save(service);
    }
}

File: frontend/src/services/serviceService.js
import { api } from './api';

export const serviceService = {
  // Get all services
  getAllServices: async () => {
    try {
      const response = await api.get('/services');
      return response.data;
    } catch (error) {
      console.error('Error fetching services:', error);
      throw error;
    }
  },

  // Get available services
  getAvailableServices: async () => {
    try {
      const response = await api.get('/services/available');
      return response.data;
    } catch (error) {
      console.error('Error fetching available services:', error);
      throw error;
    }
  },

  // Get service by ID
  getServiceById: async (serviceId) => {
    try {
      const response = await api.get(`/services/${serviceId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching service:', error);
      throw error;
    }
  },

  // Search services
  searchServices: async (keyword) => {
    try {
      const response = await api.get(`/services/search?keyword=${encodeURIComponent(keyword)}`);
      return response.data;
    } catch (error) {
      console.error('Error searching services:', error);
      throw error;
    }
  },

  // Get services by category
  getServicesByCategory: async (category) => {
    try {
      const response = await api.get(`/services/category/${encodeURIComponent(category)}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching services by category:', error);
      throw error;
    }
  },

  // Get service image URL
  getServiceImageUrl: (serviceId) => {
    return `${process.env.REACT_APP_API_BASE_URL || 'http://localhost:8081/api'}/services/${serviceId}/image`;
  },

  // Get service image as blob
  getServiceImage: async (serviceId) => {
    try {
      const response = await api.get(`/services/${serviceId}/image`, {
        responseType: 'blob'
      });
      return URL.createObjectURL(response.data);
    } catch (error) {
      console.error(`Error fetching service image ${serviceId}:`, error);
      return null;
    }
  },

  // Add service to cart
  addServiceToCart: async (serviceId, rentalPeriod, rentalPeriodType, unitPrice) => {
    try {
      const response = await api.post('/cart/add-service', {
        serviceId,
        rentalPeriod,
        rentalPeriodType,
        unitPrice
      });
      return response.data;
    } catch (error) {
      console.error('Error adding service to cart:', error);
      throw error;
    }
  },

  // Remove service from cart
  removeServiceFromCart: async (serviceId) => {
    try {
      const response = await api.delete(`/cart/remove-service/${serviceId}`);
      return response.data;
    } catch (error) {
      console.error('Error removing service from cart:', error);
      throw error;
    }
  },

  // Admin CRUD Operations
  
  // Create service
  createService: async (serviceData) => {
    try {
      const response = await api.post('/services', serviceData);
      return response.data;
    } catch (error) {
      console.error('Error creating service:', error);
      throw error;
    }
  },

  // Create service with image upload
  createServiceWithImage: async (serviceData, imageFile = null) => {
    try {
      const formData = new FormData();
      
      // Append all service fields
      formData.append('serviceName', serviceData.serviceName);
      formData.append('description', serviceData.description || '');
      formData.append('pricePerDay', serviceData.pricePerDay || serviceData.price || 0);
      formData.append('pricePerHour', serviceData.pricePerHour || 0);
      formData.append('category', serviceData.category || '');
      formData.append('isAvailable', serviceData.isAvailable || true);
      formData.append('minRentalPeriod', serviceData.minRentalPeriod || 1);
      formData.append('maxRentalPeriod', serviceData.maxRentalPeriod || 720);
      
      if (imageFile) {
        formData.append('imageFile', imageFile);
      }

      const response = await api.post('/services/with-image', formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });
      return response.data;
    } catch (error) {
      console.error('Error creating service with image:', error);
      throw error;
    }
  },

  // Update service
  updateService: async (serviceId, serviceData) => {
    try {
      const response = await api.put(`/services/${serviceId}`, serviceData);
      return response.data;
    } catch (error) {
      console.error('Error updating service:', error);
      throw error;
    }
  },

  // Delete service
  deleteService: async (serviceId) => {
    try {
      await api.delete(`/services/${serviceId}`);
    } catch (error) {
      console.error('Error deleting service:', error);
      throw error;
    }
  }
};